### PR TITLE
Add ReportingObserver 

### PR DESCRIPTION
**Pull request by RByers**
Adds initial definitions and hooks for ReportingObserver for #29.

***
<a href="https://pr-preview.s3.amazonaws.com/RByers/reporting/pull/47.html" title="Last updated on Apr 9, 2018, 5:55 PM GMT (79942d6)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/reporting/47/33808fc...RByers:79942d6.html" title="Last updated on Apr 9, 2018, 5:55 PM GMT (79942d6)">Diff</a>

### CODE REVIEW

**Comment by RByers** [comment]
@mikewest @patrickkettner @paulmeyer90 what do you think of this?  I hook into the queuing logic [here](https://rbyers.github.io/reporting/#queue-report) and invoke [this observer-specific machinery](https://rbyers.github.io/reporting/#observers).  All changes elsewhere in the spec should be unobservable I believe - they're just to resolve definition issues and make bikeshed happy.

I'm starting to work on some web platform tests - there are a surprising number of observable edge conditions to validate.  Next major step is to define deprecation reports, but I wanted to see if the basic infrastructure looked OK to you first.

File: index.src.html [hunk 1]
```diff
@@ -750,6 +756,130 @@ <h3 id="try-delivery" algorithm>
...
+callback ReportingObserverCallback = void (sequence&lt;Report> reports, ReportingObserver observer);
+
+dictionary ReportingObserverOptions {
+  sequence&lt;DOMString> types;
+};
+  </pre>
+  A {{ReportingObserver}} object can be used to observe some types of
+  <a>reports</a> from JavaScript when they are created.
+  
+  Issue: How do we specify which reports are observable?  Network-layer reports
```
Comment (igrigorik) [inline index.src.html]: In theory, that should be specified when the report is queued, as a default "off" parameter?

File: index.src.html [hunk 2]
```diff
@@ -750,6 +756,130 @@ <h3 id="try-delivery" algorithm>
...
+     list of <a>registered reporting observers</a> for the <a>environment
+     settings object</a> associated with the <a>context object</a>.
+  
+  The <dfn method for=ReportingObserver><code>disconnect()</code></dfn> method,
+  when invoked, must run these steps:
+  
+  1. If present, remove the <a>context object</a> from the list of
+     <a>registered reporting observers</a> for the <a>environment settings
+     object</a> associated with the <a>context object</a>.
+
```
Comment (igrigorik) [inline index.src.html]: Based on experience with PerformanceObserver (see https://github.com/w3c/performance-timeline/issues/74), we should also support takeRecords().

Comment (igrigorik) [reply index.src.html]: Another thing to consider: what about reports queued during "boot time" of the page, before observer can be registered? E.g. it's best practice to remove analytics and other non-critical work from the critical path, but that means that I may miss a whole lot of reports. 

For PerfObserver, we're adding support for buffering events up to a certain point: https://github.com/w3c/performance-timeline/issues/81

File: index.src.html [hunk 3]
```diff
@@ -750,6 +756,130 @@ <h3 id="try-delivery" algorithm>
...
+     1. If |observer|'s <a>options</a> has a {{ReportingObserverOptions/types}}
+        member which does not contain |report|'s [=report/type=], skip these
+        sub-steps.
+        
+     2. Create a new {{Report}} |r| with {{Report/type}} initialized to
+        |report|'s [=report/type=], {{Report/url}} initialized to |report|'s
+        [=report/url=], and {{Report/body}} initialized to |report|'s
+        [=report/body=].
+
+        Issue: how to polymorphically initialize body?
```
Comment (igrigorik) [inline index.src.html]: Isn't body a JSON payload?

File: index.src.html [hunk 4]
```diff
@@ -750,6 +756,130 @@ <h3 id="try-delivery" algorithm>
...
+       
+     3. Add |r| to the end of |observer|'s <a>report list</a>.
+         
+  2. <a>Queue a task</a> to [[#invoke-observers]] with a copy of the
```
Comment (igrigorik) [inline index.src.html]: This should probably be queued as a low priority task? For PerfObserver we define own task source and indicate that it should be a low priority queue: https://w3c.github.io/performance-timeline/#queue-a-performanceentry

File: index.src.html (hunk 1 above)
Comment (RByers) [reply index.src.html]: I think network-only reports should be the exception.  So how about a default-off "network-only" parameter?

File: index.src.html (hunk 4 above)
Comment (RByers) [reply index.src.html]: Thanks, filed https://github.com/WICG/reporting/issues/58 to discuss further.

File: index.src.html (hunk 2 above)
Comment (RByers) [reply index.src.html]: Thanks!
Discussed takeRecords as part of https://github.com/WICG/reporting/issues/58.
Filed https://github.com/WICG/reporting/issues/59 to discuss boot-time.

File: index.src.html (hunk 3 above)
Comment (RByers) [reply index.src.html]: No, it's just an object (i.e. you can do "report.body.message").  For the HTTP path, this object gets serialized to JSON.  That said @patrickkettner suggests that the text I have here is fine.  We'll have, eg. a [DeprecationReport interface](https://cs.chromium.org/chromium/src/third_party/WebKit/Source/core/frame/DeprecationReport.idl?q=DeprecationReport&sq=package:chromium&dr=C&l=10) which is an instance of [ReportBody](https://rbyers.github.io/reporting/#reportbody).  It's fine to have text elsewhere that creates a `DeprecationReport` and then manipulate it here as `ReportBody`.

**Comment by paulmeyer90** [comment]
@igrigorik I've made some changes to ReportingObserver, specifically around:

1) How observers are registered.
2) How reports are buffered before being sent to observers, and then queued in the observers before being reported.
3) The interaction between the |futureOnly| options and the report buffer.
4) Moving the setting of options from inside the observe() function to directly in the construction of the observer. Note that this prevents a number of edge cases that could arise from changing options after reports have already been observed or reported. Also, I don't think setting |futureOnly| actually makes sense after an observer is already observing, and likewise I don't see the need to change the set of report types to be observed after an observer has already been observing other types.
5) Nailed down a number of the algorithms to ensure that everything is behaving consistently, especially with respect to the lifetime of reports, and how they are buffered/queued, depending on the options.

Any thoughts?

File: index.src.html [hunk 5]
```diff
@@ -155,12 +155,12 @@ <h3 id="examples">Examples</h3>
     define a set of reporting endpoints named "`endpoint-1`":
 
     <pre>
-      <a>Report-To</a>: { "<a>url</a>": "https://example.com/reports",
-                   "<a>group</a>": "endpoint-1",
-                   "<a>max-age</a>": 10886400 },
-                 { "<a>url</a>": "https://backup.com/reports",
-                   "<a>group</a>": "endpoint-1",
-                   "<a>max-age</a>": 10886400 }
+      <a>Report-To</a>: { "<a for="ReportTo">url</a>": "https://example.com/reports",
```
Comment (igrigorik) [inline index.src.html]: As a heads up, @dcreager is working on some updates that will break this.

File: index.src.html [hunk 6]
```diff
@@ -749,6 +755,166 @@ <h3 id="try-delivery" algorithm>
...
 </section>
 
+<section>
+  <h2 id="observers">Reporting Observers</h2>
+
+  A <dfn>reporting observer</dfn> observes some types of <a>reports</a> from
+  JavaScript, and is represented in JavaScript by the {{ReportingObserver}}
+  object.
+
+  Each <a>environment settings object</a> has a <dfn>registered reporting
```
Comment (igrigorik) [inline index.src.html]: For PO we use "Each ECMAScript global environment has": https://w3c.github.io/performance-timeline/#performance-timeline

File: index.src.html [hunk 7]
```diff
@@ -749,6 +755,166 @@ <h3 id="try-delivery" algorithm>
...
+  observer list</dfn>, which is an <a>ordered set</a> of <a>reporting observers</a>.
+
+  Any <a>reporting observer</a> that is in a <a>registered reporting observer
+  list</a> is considered <dfn>registered</dfn>.
+
+  Each <a>environment settings object</a> has a <dfn>report buffer</dfn>, which
+  is an ordered list of <a>reports</a> that have been generated in
+  that <a>environment settings object</a>. This list is initially empty, and the
+  reports are stored in the same order in which they are generated.
```
Comment (igrigorik) [inline index.src.html]: For PO we queue the pending entries / reports into each observer, instead of keeping a global buffer:
https://w3c.github.io/performance-timeline/#queue-a-performanceentry

File: index.src.html [hunk 8]
```diff
@@ -749,6 +755,166 @@ <h3 id="try-delivery" algorithm>
...
+
+  <h3 id=interface-reporting-observer>Interface {{ReportingObserver}}</h3>
+
+  <pre class="idl">
+[NoInterfaceObject]
```
Comment (igrigorik) [inline index.src.html]: The body is a JS object, right? Should this really be NoInterfaceObject?

File: index.src.html [hunk 9]
```diff
@@ -749,6 +755,166 @@ <h3 id="try-delivery" algorithm>
...
+interface Report {
+  readonly attribute DOMString type;
+  readonly attribute DOMString url;
+  readonly attribute ReportBody? body;
+};
+
+[Constructor(ReportingObserverCallback callback, optional ReportingObserverOptions options)]
+interface ReportingObserver {
+  void observe();
+  void disconnect();
```
Comment (igrigorik) [inline index.src.html]: We should also define takeRecords for consistency — you'll need it. See https://github.com/w3c/performance-timeline/commit/aecfdf45547855df9d4d7d33d30aaa9ffc1a9781.

File: index.src.html [hunk 10]
```diff
@@ -749,6 +755,166 @@ <h3 id="try-delivery" algorithm>
...
+};
+
+callback ReportingObserverCallback = void (sequence&lt;Report> reports, ReportingObserver observer);
```
Comment (igrigorik) [inline index.src.html]: `&lt;`?

File: index.src.html [hunk 11]
```diff
@@ -749,6 +755,166 @@ <h3 id="try-delivery" algorithm>
...
+
+dictionary ReportingObserverOptions {
+  sequence&lt;DOMString> types;
+  boolean futureOnly = false;
```
Comment (igrigorik) [inline index.src.html]: For sake of consistency, s/futureOnly/buffered? 

https://w3c.github.io/performance-timeline/#performanceobserverinit-dictionary

File: index.src.html [hunk 12]
```diff
@@ -749,6 +755,166 @@ <h3 id="try-delivery" algorithm>
...
+};
+  </pre>
+
+  Issue: How do we specify which reports are observable?  Network-layer reports
```
Comment (igrigorik) [inline index.src.html]: It seems that each report type should be explicitly stating if it's meant to be JS-observable. We can make this a flag on the ~queue the report step, with default value set to "false".

File: index.src.html [hunk 13]
```diff
@@ -749,6 +755,166 @@ <h3 id="try-delivery" algorithm>
...
+     the <a>context object</a>.
+
+  2. Append the <a>context object</a> to the <a>registered reporting observer
+     list</a> of its associated <a>environment settings object</a>.
+
+  3. If the <a>context object</a>'s {{futureOnly}} <a>option</a> is true, return.
+
+  4. Set <a>context object</a>'s {{futureOnly}} <a>option</a> to true.
+
+  5. For each |report| in the <a>report buffer</a> associated with |settings|,
```
Comment (igrigorik) [inline index.src.html]: This implies that we must queue all reports indefinitely for the lifetime of the page. Is that the behavior we want, and/or reason for concern? It's not crazy to imagine hundreds of reports sitting around indefinitely..

File: index.src.html [hunk 14]
```diff
@@ -749,6 +755,166 @@ <h3 id="try-delivery" algorithm>
...
+
+  1. Create a new {{ReportingObserver}} object |observer|.
+
+  2. Set |observer|'s <a>callback</a> to |callback|.
+
+  3. Set |observer|'s <a>options</a> to |options|.
+
+  4. Return |observer|.
+
+  The <dfn method for=ReportingObserver><code>observe()</code></dfn>
```
Comment (igrigorik) [inline index.src.html]: This definition doesn't cover the case where I call observe() on an existing observer with different options. 

See https://w3c.github.io/performance-timeline/#the-performanceobserver-interface for an example.

File: index.src.html [hunk 15]
```diff
@@ -749,6 +755,166 @@ <h3 id="try-delivery" algorithm>
...
+  This algorithm makes |report|'s contents available to any <a>registered</a>
+  <a>reporting observers</a> on the provided <a>environment settings object</a>
+  |settings|.
+
+  1. For each {{ReportingObserver}} |observer| <a>registered</a> with
+     |settings|, execute [[#add-report]] on |report| and |observer|.
+
+  2. Add |report| to the <a>report buffer</a> associated with |settings|.
+
+  3. If the <a>report buffer</a> now contains more than 100 reports, remove the
```
Comment (igrigorik) [inline index.src.html]: This implies there is a global 100 reports limit? 

What is the "beginning"? We should spell it out.

**Comment by igrigorik** [comment]
Paging @dcreager for review :)

File: index.src.html (hunk 6 above)
Comment (paulmeyer90) [reply index.src.html]: Okay, I'll use "ECMAScript global environment" instead.

File: index.src.html (hunk 7 above)
Comment (paulmeyer90) [reply index.src.html]: This buffer is just used for the purpose of passing past reports into new observers. Pending reports are still queued on observers (in their "report list").

File: index.src.html (hunk 11 above)
Comment (paulmeyer90) [reply index.src.html]: Done.

File: index.src.html (hunk 8 above)
Comment (paulmeyer90) [reply index.src.html]: Yes, I'll remove those tags.

File: index.src.html (hunk 13 above)
Comment (paulmeyer90) [reply index.src.html]: This is so that observers can be made to observe reports that were generated before they could be registered, and this is the behavior we want. We don't expect that the number of reports will be that large, but there is also the buffer limit of 100 in place, so there can't be hundreds sitting around.

File: index.src.html (hunk 14 above)
Comment (paulmeyer90) [reply index.src.html]: That issue is not present here because this spec no longer allows options to be passed into observe() for ReportingObserver. The options are passed in on construction of the observer, and so cannot be changed with multiple calls to observe(), for example.

File: index.src.html (hunk 15 above)
Comment (paulmeyer90) [reply index.src.html]: Yes, there is a 100 report buffer limit.

The "beginning" of the list is used here: https://infra.spec.whatwg.org/#lists, so I think it is unambiguous. If it does need specifying though, then it should probably be specified in infra rather than in this spec.

File: index.src.html (hunk 9 above)
Comment (paulmeyer90) [reply index.src.html]: Okay, added.

File: index.src.html (hunk 12 above)
Comment (paulmeyer90) [reply index.src.html]: Okay, I've added a step to check for this. Each specific report type will need to explicitly state that it is observable from JavaScript in order to be observed by a ReportingObserver.

**Comment by paulmeyer90** [comment]
The formatting changes were actually necessary in order to allow bikeshed to properly link everything with the new changes, and so cannot be separated out from this CL.

File: index.src.html (hunk 13 above)
Comment (dcreager) [reply index.src.html]: There's a section further down about garbage collection, which gives the user agent the ability to throw away reports before they're delivered if they've been sitting around too long.  I suggest updating that text to say that if any reports are ever GCed from the reporting cache, they must also be removed from any observer's report list that they appear in.

File: index.src.html (hunk 7 above)
Comment (dcreager) [reply index.src.html]: Is this to handle a race condition, where an observer might be registered after some of the events that it needs to monitor?  If so I suggest some explanatory text somewhere; the name `buffered` implied to me something about queueing up reports between when Reporting receives from the downstream spec and before they're handed over to the observer.

File: index.src.html (hunk 12 above)
Comment (dcreager) [reply index.src.html]: The parameter would need to be per-report, not per-type.  Only **some** Network Error Logging reports should be visible to observers, for instance.  In particular, a NEL report should only be visible to the same origin that the report is about.  So for instance, if a page at `foo.com` contains a link to `bar.com`, `bar.com` can request NEL reports, but we **won't** send those reports about `bar.com` to any `foo.com` Reporting collectors.  And so any Reporting observers registered for `foo.com` shouldn't be able to see those `bar.com` NEL reports, either.

I **think** we can still get that with the parameter that the downstream spec has to set, indicating whether individual reports are visible to observers or not.  NEL should have all the info needed to fill that in — if the environment settings object's origin is different than the origin of the request that the report describes, set it to `no`; otherwise set it to `yes`.  I'd appreciate someone checking my work on that, though.  :wink: 

So in addition to the parameter being present, I'd suggest some clarifying text.  ("Downstream specs should make sure to set this parameter so that observers are only able to see reports that would be sent to collectors for that origin" or something like that)

**Comment by paulmeyer90** [comment]
@dcreager: Not sure why I can't reply to your comments directly. Anyways:

1. I have added a note to the garbage collection section as you described.

2. I have added a note to describe the purpose of the |report buffer| and |buffered| option.

3. Currently, the spec is set up to not allow any reports to be observable cross-origin; they are only sent to observers in the same environment settings object. I think that this property, in addition to specifying which report types are observable, should be sufficient. Are there any other problem cases you can think of that would require observability to be specified per-report rather than per-type? I have also added a note to clarify this in the spec.

File: index.src.html [hunk 16]
```diff
@@ -769,6 +788,191 @@ <h3 id="try-delivery" algorithm>
...
+  void observe();
+  void disconnect();
+  ReportList takeRecords();
+};
+
+callback ReportingObserverCallback = void (sequence&lt;Report> reports, ReportingObserver observer);
+
+dictionary ReportingObserverOptions {
+  sequence&lt;DOMString> types;
+  boolean buffered = true;
```
Comment (dcreager) [inline index.src.html]: Can we change this to `processAlreadyBufferedReports` (or something in that vein)?  Unless there's a general preference in these specs for shorter names, I like the more descriptive style, especially for boolean options.

Comment (igrigorik) [reply index.src.html]: I'd prefer current name because it allows us to keep consistency between *Observer interfaces — e.g. PerformanceObserver, ReportingObserver, etc.

**Comment by igrigorik** [comment]
> @igrigorik, is there already boilerplate text about that that we can steal from other specs? Or is it implicit and not worth mentioning?

I'm not aware of boilerplate, but you may want to add a note (non-normative, probably) that outlines how non UA clients should behave.

File: index.src.html (hunk 16 above)
Comment (dcreager) [reply index.src.html]: SGTM
